### PR TITLE
Retry unauthorized HTTP request with a renewed JWT token

### PR DESCRIPTION
**Pull request by didier-wenzek**
Signed-off-by: Didier Wenzek <didier.wenzek@free.fr>

## Proposed changes

Address the two root causes of [Unauthorized request](https://github.com/thin-edge/thin-edge.io/issues/1562):i

* Re-open the MQTT connection for each JWT token request.
* When an HTTP request to Cumulocity is rejected as unauthorized, then renew the JWT token and retry.
Do this only once, since there is no point to get yet another fresh token if rejected twice in a row.

For the first point, the alternative is to keep the MQTT connection open and to have a background task that consumes all the JWT token - requested by the process or not, and to cache the latest one. However, this alternative introduces more complexity and opening an MQTT connection before each HTTP request should not be a performance issue for the current use-cases.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

See https://github.com/thin-edge/thin-edge.io/issues/1562

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/c8y_api/src/http_proxy.rs [hunk 1]
```diff
@@ -621,6 +632,18 @@ mod tests {
         let config_file = create_test_config_file_with_content(config_content)?;
 
         // Mock endpoint for config upload event creation
+        // Fake an expired JWT token by rejecting the first config upload request
```
Comment (albinsuresh) [inline crates/core/c8y_api/src/http_proxy.rs:635]: Isn't it better to have a dedicated test for this retry feature?

**Review by albinsuresh — COMMENTED** [review]
The `try_get_internal_id` function also needs to be updated with this retry logic.

**Review by reubenmiller — APPROVED** [review]
Looks good.

Moving forward we probably need a more generic retry handler to also handle failed requests due to network interruptions etc. But it is not for this ticket's scope as there would be additional aspects like backing off before retrying etc.

**Review by reubenmiller — COMMENTED** [review]
Looks good.

Moving forward we probably need a more generic retry handler to also handle failed requests due to network interruptions etc. But it is not for this ticket's scope as there would be additional aspects like backing off before retrying etc.

File: crates/core/c8y_api/src/http_proxy.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/c8y_api/src/http_proxy.rs:631]: I agree that I have been a bit lazy here ;-) Will see if I find some time for that.

**Comment by didier-wenzek** [comment]
> The `try_get_internal_id` function also needs to be updated with this retry logic.

Done with https://github.com/thin-edge/thin-edge.io/pull/1586/commits/626d934f69e09977758500109af03f0f4a346708

**Review by albinsuresh — APPROVED** [review]
Neat trick using a fresh connection for every token request. The fix wasn't obvious to me on a first look, as I was searching for the logic that iterates over the message stream for the most recent token.

**Review by reubenmiller — CHANGES_REQUESTED** [review]
One of the checks is failing due to a `fixup!` commit, so just so we following the correct process could you please squash the changes to remove the `fixup!` commit?

**Comment by didier-wenzek** [comment]
> Neat trick using a fresh connection for every token request. The fix wasn't obvious to me on a first look, as I was searching for the logic that iterates over the message stream for the most recent token.

I found no crystal clear way to get the most recent token. The main issue is that consuming messages from a subscription there is no way to know that we have read all the available messages.

**Comment by reubenmiller** [comment]
> > One of the checks is failing due to a `fixup!` commit, so just so we following the correct process could you please squash the changes to remove the `fixup!` commit?
> 
> Sure, I will `git rebase -i --autosquash main`. We use to do that only once the PR approved, the goal being to avoid `push --force` during the review process.

Ok, thanks for the clarification. Approved

**Review by reubenmiller — APPROVED** [review]
Looks good

File: crates/core/c8y_api/src/http_proxy.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/c8y_api/src/http_proxy.rs]: Added with https://github.com/thin-edge/thin-edge.io/pull/1586/commits/73d114f708f378f630804588f36919af788f4e97

**Comment by albinsuresh** [comment]
> > Neat trick using a fresh connection for every token request. The fix wasn't obvious to me on a first look, as I was searching for the logic that iterates over the message stream for the most recent token.
> 
> I found no crystal clear way to get the most recent token. The main issue is that consuming messages from a subscription there is no way to know that we have read all the available messages.

This may have been an option: https://docs.rs/futures/0.3.25/futures/future/trait.FutureExt.html#method.now_or_never by repeatedly calling it until it returns `None`.

**Comment by didier-wenzek** [comment]
> > > Neat trick using a fresh connection for every token request. The fix wasn't obvious to me on a first look, as I was searching for the logic that iterates over the message stream for the most recent token.
> > 
> > 
> > I found no crystal clear way to get the most recent token. The main issue is that consuming messages from a subscription there is no way to know that we have read all the available messages.
> 
> This may have been an option: https://docs.rs/futures/0.3.25/futures/future/trait.FutureExt.html#method.now_or_never by repeatedly calling it until it returns `None`.

Thanks for the hint. Indeed this might be useful. However, in that specific case, complexity arises due to a race condition between this call and the reception of a token. One might get `None` not because all the published token have been read, but because the latest taken as not been received yet.

**Pull request merged**